WordPress/gutenberg
# Site Editor: Add ability to focus on editing a page's content vs the page's template

## What?
Second attempt at https://github.com/WordPress/gutenberg/pull/49980. The key difference is that it's based on https://github.com/WordPress/gutenberg/pull/50643 instead of https://github.com/WordPress/gutenberg/pull/50082.

Requires https://github.com/WordPress/gutenberg/pull/50643.

Part of https://github.com/WordPress/gutenberg/issues/49404 and https://github.com/WordPress/gutenberg/issues/44461.

Adds functionality to the Site Editor that lets you, when editing a page, switch between focusing on editing the page's content versus editing the page's template.

## Why?
Adding some separation helps to prevent a very common point of frustration encountered when a user attempts to change something on a specific page but the change is made everywhere.

## How?
- Add breadcrumb and sidebar UI to `wp-edit-site` that lets the user toggle between focusing on editing the template and focusing on editing the page.
- Use the new `useBlockEditingMode()` hook added in https://github.com/WordPress/gutenberg/pull/50643 to prevent editing blocks that aren't the Post Title, Post Featured Image and Post Content block.
- Various UI affordances.
  - Notifications to nudge the user to edit the template when they click on a locked block.
  - A custom right sidebar for when the focus is on editing the page.

## Testing Instructions
1. Ensure that you have a primary Navigation menu saved.
1. Go to _Appearance → Editor → Navigation_ and select a page.
1. Test the new switching functionality.
1. Check that navigating to _Appearance → Editor → Templates_ and selecting a template results in the same site editor experience as before (no switching functionality).

## Screenshots or screencast 

https://user-images.githubusercontent.com/612155/233555590-9036ad7c-e3ad-47dd-8861-8058bb7168d4.mp4

## Follow up tasks

- [x] Rename all the methods I'm adding to refer to `post` instead of `page`. 
- [x] DRY up `ContentBlocksList` and `BlockInspectorLockedBlocks`.
- [ ] Explore opportunities for user education e.g. display a guide when switching to the template.
- [x] Make Post Title, Post Featured Image, Post Content blocks say "Page" instead of "Post" https://github.com/WordPress/gutenberg/issues/48963
- [ ] Explore making `useBlockEditingMode()` available for extenders. (Mark it public, add docs, add dev note.)
- [ ] A black pixel appears where the toolbar used to appear when Post Title is selected. 
- [ ] Try to remove `__unstable` `contentOnly` methods. (Not really related but using this list as my todo list now 😂)
- [ ] Make the notifications appear in the top centre with a purple background. As in the design mockup.
- [ ]  [Make it so that the global inserter inserts content into the Post Content block when no block is selected.](https://github.com/WordPress/gutenberg/issues/51538#issuecomment-1593097498)
- [x] Remove `select()` hack from `getBlockEditingMode` when https://github.com/WordPress/gutenberg/pull/50985 is fixed.
- [ ] Optimise selectors https://github.com/WordPress/gutenberg/pull/50643#pullrequestreview-1450920100 
- [ ] Add an E2E test
- [ ] https://github.com/WordPress/gutenberg/issues/51259
- [ ] Add 'Edit template' and 'Back to page' commands to the command centre
